src/Raven.Client/Documents/Session/AsyncSessionDocumentTimeSeries.cs

```diff
@@ -24,13 +26,56 @@ public AsyncSessionDocumentTimeSeries(InMemoryDocumentSessionOperations session,
         }
         public async Task<IEnumerable<TimeSeriesValue>> GetAsync(string timeseries, DateTime from, DateTime to, CancellationToken token = default)
         {
-            if (Session.TimeSeriesByDocId.TryGetValue(DocId, out var cache))
+            TimeSeriesDetails details;
+            
+            if (Session.TimeSeriesByDocId.TryGetValue(DocId, out var cache) &&
+                cache.TryGetValue(timeseries, out var ranges))
             {
-                if (cache.Values.TryGetValue(timeseries, out var series))
+                if (ranges[0].From > to || ranges[ranges.Count - 1].To < from)
+                {
+                    // the entire range [from, to] is out of cache bounds
+
+                    Session.IncrementRequestCount();
+
+                    details = await Session.Operations.SendAsync(
+                            new GetTimeSeriesOperation(DocId, timeseries, from, to), Session.SessionInfo, token: token)
+                        .ConfigureAwait(false);
+
+                    if (Session.NoTracking == false)
+                    {
+                        var index = ranges[0].From > to ? 0 : ranges.Count;
+                        ranges.Insert(index, details.Values[timeseries][0]);
+                    }
+
+                    return details.Values[timeseries][0].Values;
+                }
+
+                if (GetFromCacheOrFigureOutMissingParts(from, to, ranges,
+                    out var fromRangeIndex, out var toRangeIndex,
+                    out var rangesToGetFromServer, out var resultToUser))
                 {
-                    //TODO: we need to properly handle this here, including merging multiple sections, getting just the values we need, etc.
-                    return series.Values;
+                    return resultToUser;
                 }
+
+                // can't get the entire range from cache
+                Session.IncrementRequestCount();
+
+                details = await Session.Operations.SendAsync(
+                        new GetTimeSeriesOperation(DocId, timeseries, rangesToGetFromServer), Session.SessionInfo, token: token)
+                    .ConfigureAwait(false);
+
+                var values = MergeRangesAndResults(
+                    fromRangeIndex, 
+                    toRangeIndex == ranges.Count ? ranges.Count - 1 : toRangeIndex,
```

Comment:
Given that this is only used here, can't we already modify that in the `GetFromCacheOrFigureOutMissingParts` and spare this logic?

Reply:
done